File: Episodio04/Ruby/Ejemplo-Optional/customer.rb
require 'rumonade'

class Customer

  INVALID_NAME = 'Name can not be empty'

  @name
  @address

  def self.named(name,address)
    raise Exception.new INVALID_NAME if name.empty?

    new name, Option(address)
  end

  def initialize(name,address)
    @name = name
    @address = address
  end

  def address
    @address
  end
end

File: Temporada01/Episodio04 - No usar null/Java/Episodio04-Ejemplo-Optional/src/Customer.java
import java.util.Optional;

public class Customer {

	public static final String INVALID_NAME = "Name can not be empty";
	
	private String name;
	private Optional<Address> address;

	public static Customer named(String name, Address address) {
		if(name.isEmpty()) throw new RuntimeException(INVALID_NAME);
		
		return new Customer(name, Optional.ofNullable(address));
	}

	private Customer(String name, Optional<Address> address) {
		this.name = name;
		this.address = address;
	}

	public Optional<Address> getAddress() {
		return address;
	}
	
}
